### PR TITLE
Add S3 support for graphs

### LINKED ISSUE #5: Upload image with guid name to S3 before posting to hubot
To make this work with all clients of Hipchat and Slack the image needs to be publicly accessible, most Grafana installs will be behind firewalls/dmz so the best option is to upload the image to an s3 bucket and link to that.

### PR DESCRIPTION
**Pull request by stephenyeargin**
Prior to this PR, a user would have to meet several requirements before the graphs would render on services such as Slack or HipChat. First, they would have to have a wide-open Grafana instance (no authentication). Second, it would have to be on the public internet (accessible from the hosted chat's servers).

This PR handles both of these instances by allowing the user to configure an S3 bucket and credentials. With those credentials, the script will download the image using authentication headers and upload the chart image to the defined S3 path. The script will then provide a link per usual, which will be rendered on the hosted chat service.

Special thanks to @rick, who I (mostly|poorly) ripped off this entire implementation from his `hubot-graphme` package. :grinning:

Fixes #5 /cc @torkelo @tenzer

### CODE REVIEW

**Comment by torkelo** [comment]
:+1:

File: src/grafana.coffee [hunk 1]
```diff
@@ -154,3 +184,49 @@ module.exports = (robot) ->
...
+
+    filename = uploadPath()
+
+    req = client.put(filename, headers)
+    req.on 'response', (res) ->
+      if (200 == res.statusCode)
+        image = "https://s3.amazonaws.com/#{s3_bucket}/#{filename}"
+        sendRobotResponse msg, title, image, link
+      else
+        msg.send "#{title} - [Upload Error] - #{link}"
```
Comment (Tenzer) [inline src/grafana.coffee]: Perhaps it would be a good idea to log any error we get back in the Hubot log?

File: src/grafana.coffee [hunk 2]
```diff
@@ -154,3 +184,49 @@ module.exports = (robot) ->
...
+      key    : s3_access_key
+      secret : s3_secret_key,
+      bucket : s3_bucket
+    }
+
+    headers = {
+      'Content-Length' : length,
+      'Content-Type'   : content_type,
+      'x-amz-acl'      : 'public-read',
+      "encoding"       : null
```
Comment (Tenzer) [inline src/grafana.coffee]: You have some mixed quotes here, it doesn't really matter much, it just looks a bit odd :)

File: src/grafana.coffee [hunk 3]
```diff
@@ -154,3 +184,49 @@ module.exports = (robot) ->
...
       data = JSON.parse(body)
       return callback(data)
+
+  # Pick a random filename
+  uploadPath = () ->
+    prefix = s3_prefix || 'grafana'
+    "#{prefix}/#{crypto.randomBytes(20).toString('hex')}.png"
+
+  # Fetch an image from provided URL, upload it to S3, returning the resulting URL
+  fetchAndUpload = (msg, title, url, link, callback) ->
```
Comment (Tenzer) [inline src/grafana.coffee]: The callback isn't called at any time, is that for future use cases? If so, I think a call to the callback function should be added to the function so it actually works if we want to make use of it at some point.
If the callback parameter is removed, then it should also be removed further up where `fetchAndUpload()` is called.

File: src/grafana.coffee (hunk 1 above)
Comment (Tenzer) [inline src/grafana.coffee]: I can see the problem of `uploadToS3()` not having any reference to `robot` and the logger, but perhaps `console.log()` could work? It seems like `res.statusMessage` is the interesting bit to print.

**Comment by Tenzer** [comment]
I have added a few comments to this PR and made #16 which should be possible to merge into this one, which adds support for S3 buckets in order regions than the US.

File: src/grafana.coffee (hunk 1 above)
Comment (stephenyeargin) [inline src/grafana.coffee]: I'm pretty sure the `robot` object is available here -- I just forgot to add it.

File: src/grafana.coffee (hunk 2 above)
Comment (stephenyeargin) [inline src/grafana.coffee]: Yeah, that's copy/:spaghetti: from the other package. I need to run the whole file through a style check.

File: src/grafana.coffee (hunk 3 above)
Comment (stephenyeargin) [inline src/grafana.coffee]: Yeah, I had originally tried to have yet another callback here, but the execution order got out of whack. There's some flow work that could be done here to a) use less code and b) make it more readable. I think I'll ditch the callback parameter for now and refactor it as time goes on.

File: src/grafana.coffee (hunk 2 above)
Comment (rick) [inline src/grafana.coffee]: :trollface:

**Comment by stephenyeargin** [comment]
Alright, I think that takes care of it.

**Pull request merged**